### PR TITLE
txhasher: adding a txhasher calss for different Arbitrum specific tx types

### PR DESCRIPTION
**Pull request by mahsamoosavi**
The previous [PR](https://github.com/OffchainLabs/arbitrum-sdk/pull/165) had some unrelated changes from another PR.
Had to force push backwards and create a new pull request. I'll be applying all those changes @yahgwai has requested on the previous one!

### CODE REVIEW

File: src/lib/message/txHashEncoder.ts [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+   * @param type //type of the tx
+   * @param l2ChainId
+   * @param fromAddress
+   * @param messageNumber
+   * @param destAddress
+   * @param l1Value
+   * @returns
+   */
+  public static calculateArbitrumDepositTxHash(
+    type: number,
```
Comment (mahsamoosavi) [inline src/lib/message/txHashEncoder.ts:19]: @yahgwai: About [this comment](https://github.com/OffchainLabs/arbitrum-sdk/pull/165#discussion_r964618035), the reason for passing the type is to be able to double check if we're using the right function for encoding. It should be already clear based on the function names but it adds a safety check. 

I can def remove it if you think it's redundant though

Comment (mahsamoosavi) [reply src/lib/message/txHashEncoder.ts]: nevermind this ^ it's totally restructured now

File: src/lib/message/ArbTxHasher.ts [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+      case ArbTxType.Contract:
+        return '0x66'
+      case ArbTxType.Retry:
+        return '0x68'
+      case ArbTxType.SubmitRetryable:
+        return '0x69'
+      case ArbTxType.Internal:
+        return '0x6A'
+    }
+  }
```
Comment (gzeoneth) [inline src/lib/message/ArbTxHasher.ts:122]: can be 
```
`0x${txType.toString(16)}`
```

File: src/lib/message/ArbTxHasher.ts [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+   * @returns
+   */
+  public static hash(
+    tx:
+      | DepositTx
+      | UnsignedTx
+      | ContractTx
+      | RetryTx
+      | SubmitRetryableTx
+      | InternalTx
```
Comment (gzeoneth) [inline src/lib/message/ArbTxHasher.ts:136]: I think you can have each of the type extend a base type (e.g. with the type field only) so this can be cleaner.

File: src/lib/message/ArbTxHasher.ts [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+      case ArbTxType.Retry:
+        this.hashRetryTx(tx)
+        break
+      case ArbTxType.SubmitRetryable:
+        this.hashSubmitRetryableTx(tx)
+        break
+      case ArbTxType.Internal:
+        this.hashInternalTx(tx)
+        break
+    }
```
Comment (gzeoneth) [inline src/lib/message/ArbTxHasher.ts:157]: Can add 
```
    default:
      const exhaustiveCheck: never = c;
      throw new Error(`Unhandled case: ${exhaustiveCheck}`);
```
as an exhaustive check https://stackoverflow.com/a/58009992

File: src/lib/message/L1ToL2Message.ts [hunk 5]
```diff
@@ -461,17 +485,28 @@ export class L1ToL2MessageReader extends L1ToL2Message {
...
+      if (confirmations || chosenTimeout) {
+        throw new ArbSdkError(
+          `Timed out waiting to retrieve retryable creation receipt: ${this.retryableCreationId}.`
+        )
+      } else {
+        throw new ArbSdkError(
+          `Retryable creation receipt not found ${this.retryableCreationId}.`
+        )
+      }
+    }
```
Comment (gzeoneth) [inline src/lib/message/L1ToL2Message.ts:509]: these unrelated change are still here

**Review by gzeoneth — CHANGES_REQUESTED** [review]
also missing test cases

File: src/lib/message/ArbTxHasher.ts (hunk 4 above)
Comment (yahgwai) [reply src/lib/message/ArbTxHasher.ts:157]: Yeah, I like this way of doing it with an UnreachableCaseError: https://stackoverflow.com/a/52913382

File: src/lib/message/ArbTxHasher.ts [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+      | DepositTx
+      | UnsignedTx
+      | ContractTx
+      | RetryTx
+      | SubmitRetryableTx
+      | InternalTx
+  ) {
+    switch (tx.type) {
+      case ArbTxType.Deposit:
+        this.hashDepositTx(tx)
```
Comment (yahgwai) [inline src/lib/message/ArbTxHasher.ts:140]: I think these are missing the return statements - once you add that you also dont need the `break`

File: src/lib/message/ArbTxHasher.ts [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+        return '0x6A'
+    }
+  }
+
+  /**
+   * Accept the Arbitrum specific tx type and call the right function for creating the custom hash.
+   * @param ArbTxType
+   * @returns
+   */
+  public static hash(
```
Comment (yahgwai) [inline src/lib/message/ArbTxHasher.ts:129]: We still need to add tests for this func

File: src/lib/message/L1ToL2Message.ts (hunk 5 above)
Comment (mahsamoosavi) [reply src/lib/message/L1ToL2Message.ts:509]: I did reset to c-nitro-stable and force push 
Not sure why they're still there tho

File: src/lib/message/ArbTxHasher.ts (hunk 7 above)
Comment (mahsamoosavi) [reply src/lib/message/ArbTxHasher.ts]: Adding tests now

File: package.json [hunk 8]
```diff
@@ -84,7 +84,8 @@
     "typedoc": "^0.21.5",
     "typedoc-plugin-markdown": "^3.10.4",
     "typescript": "^4.2.2",
-    "yargs": "^17.3.1"
+    "yargs": "^17.3.1",
+    "ts-essentials": "^9.3.0"
```
Comment (fionnachan) [inline package.json:88]: is this package needed?

Comment (fionnachan) [reply package.json]: thanks! can you please undo the yarn.lock changes too?

Comment (mahsamoosavi) [reply package.json]: I did remove the yarn.lock file followed w a yarn install to create a new one
That should do it right?

Comment (fionnachan) [reply package.json]: i think the best way is to do the following:
```bash
git pull
git merge origin/main
```
then resolve the merge conflicts on the files using your IDE
for `yarn.lock`, accept all incoming changes for the whole file then save
stage the changes
complete the merge commit
`git push`

**Comment by yahgwai** [comment]
We should also switch the base branch of this to `main`
